library/alloc/src/collections/btree/map/entry.rs

```diff
@@ -312,7 +313,20 @@ impl<'a, K: Ord, V> VacantEntry<'a, K, V> {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn insert(self, value: V) -> &'a mut V {
-        let out_ptr = match self.handle.insert_recursing(self.key, value) {
+        let handle = match self.handle {
+            Some(handle) => handle,
+            None => {
+                // SAFETY: We have consumed self.handle and the reference returned.
+                let map = unsafe { self.dormant_map.awaken() };
+                let root = map.root.insert(NodeRef::new_leaf().forget_type());
+                // SAFETY: root is a leaf node we have just created.
+                let mut root = unsafe { root.borrow_mut().cast_to_leaf_unchecked() };
+                root.push(self.key, value);
+                map.length = 1;
+                return root.last_kv().into_val_mut();
```

Comment:
Interesting plot twist… I didn't realize you could avoid the `*mut V`, and the unsafe cast here is easily verified. Not sure which one I like better now. But here surely `NodeRef::new_leaf().forget_type()` can just be `NodeRef::new()`.

Reply:
> I didn't realize you could avoid the *mut V

Yes, but I double checked `.last_kv()` and found it involves an unnecessary check `assert!(len > 0);`, so I think the previous version (create a `Leaf` node, let push return `&mut V` to obtain a `*mut V`, and `.forget_type()` then put into the `root`) have better might performance?

Reply:
The compiler might possibly be able to wipe out that assertion. But even then, that single test is peanuts compared to avoiding an allocation. So I think the choice should be based on code readability. Right now, I slightly favour the previous form (without the block) because the cast hack is present and deeper in the other branches anyway.

Reply:
> Right now, I slightly favour the previous form (without the block) because the cast hack is present and deeper in the other branches anyway.

Sure, it is neat